### PR TITLE
Feature/publish point cloud

### PR DESCRIPTION
**Pull request by gocarlos**
Added a service to publish the intern mesh as a point cloud in a ROS topic.

### CODE REVIEW

File: InfiniTAM/Engine/RosEngine.cpp [hunk 1]
```diff
@@ -25,19 +25,28 @@ RosEngine::RosEngine(ros::NodeHandle& nh, const char*& calibration_filename)
...
+                        "/world");
   depth_info_sub =
       nh.subscribe(depth_camera_info_topic_, 1,
                    &RosEngine::depthCameraInfoCallback, (RosEngine*)this);
   rgb_info_sub =
       nh.subscribe(rgb_camera_info_topic_, 1, &RosEngine::rgbCameraInfoCallback,
                    (RosEngine*)this);
+
+  complete_point_cloud_pub_ = nh.advertise<sensor_msgs::PointCloud2>(
+      "/complete_table_top_scene_topic", 1);
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: create some parameter for this and default to something generic.

File: InfiniTAM/Engine/RosEngine.cpp [hunk 2]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
 Vector2i RosEngine::getDepthImageSize(void) { return image_size_depth_; }
 Vector2i RosEngine::getRGBImageSize(void) { return image_size_rgb_; }
 
+sensor_msgs::PointCloud2 RosEngine::conversionToPCL() {
+  ROS_INFO("conversionToPCL() start");
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: remove

File: InfiniTAM/Engine/RosEngine.cpp [hunk 3]
```diff
@@ -25,19 +25,28 @@ RosEngine::RosEngine(ros::NodeHandle& nh, const char*& calibration_filename)
...
+
   while (!rgb_info_ready_ || !depth_info_ready_) {
     ROS_INFO("Spinning, waiting for rgb and depth camera info messages.");
     ros::spinOnce();
     ros::Duration(1.0).sleep();
   }
 
+  // initialize service
+  publish_scene_service_ = nh.advertiseService(
+      "publish_scene", &RosEngine::PublishMap, (RosEngine*)this);
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: use `static_cast`

File: InfiniTAM/Engine/RosEngine.cpp [hunk 4]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
 Vector2i RosEngine::getDepthImageSize(void) { return image_size_depth_; }
 Vector2i RosEngine::getRGBImageSize(void) { return image_size_rgb_; }
 
+sensor_msgs::PointCloud2 RosEngine::conversionToPCL() {
+  ROS_INFO("conversionToPCL() start");
+  pcl::PointCloud<pcl::PointXYZ> complete_point_cloud;
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: `point_cloud_pcl`?

File: InfiniTAM/Engine/RosEngine.cpp [hunk 5]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+  sensor_msgs::PointCloud2 complete_point_cloud2;
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: rename variable to `point_cloud_msg`

File: InfiniTAM/Engine/RosEngine.cpp [hunk 6]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+
+  if (main_engine_ != NULL) {
+
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: rm unnecessary vspaces.

File: InfiniTAM/Engine/RosEngine.cpp [hunk 7]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
 Vector2i RosEngine::getDepthImageSize(void) { return image_size_depth_; }
 Vector2i RosEngine::getRGBImageSize(void) { return image_size_rgb_; }
 
+sensor_msgs::PointCloud2 RosEngine::conversionToPCL() {
+  ROS_INFO("conversionToPCL() start");
+  pcl::PointCloud<pcl::PointXYZ> complete_point_cloud;
+  sensor_msgs::PointCloud2 complete_point_cloud2;
+
+  if (main_engine_ != NULL) {
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: CHECK_NOTNULL(main_engine_)?

File: InfiniTAM/Engine/RosEngine.cpp [hunk 8]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+
+    CHECK_NOTNULL(main_engine_->GetMesh());
+
+    main_engine_->GetMeshingEngine()->MeshScene(main_engine_->GetMesh(),
+                                           main_engine_->GetScene());
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: fix format

File: InfiniTAM/Engine/RosEngine.cpp [hunk 9]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+
+    ORUtils::MemoryBlock<ITMMesh::Triangle>* cpu_triangles;
+    bool shoulDelete = false;
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: fix variable name. (and also be more descriptive)

File: InfiniTAM/Engine/RosEngine.cpp [hunk 10]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+    } else {
+      cpu_triangles = main_engine_->GetMesh()->triangles;
+    }
+
+    ITMMesh::Triangle* triangleArray = cpu_triangles->GetData(MEMORYDEVICE_CPU);
+
+    complete_point_cloud.width =
+        main_engine_->GetMesh()->noTotalTriangles * 3; // Point cloud has 3 points per triangle.
+    complete_point_cloud.height =
+        1;  // TODO(gocarlos) how to convert to structured point cloud?
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: you can't (the cloud is generally generated from multiple viewpoints)

File: InfiniTAM/Engine/RosEngine.cpp [hunk 11]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+          ORUtils::MemoryBlock<ITMMesh::Triangle>::CUDA_TO_CPU);
+      shoulDelete = true;
+    } else {
+      cpu_triangles = main_engine_->GetMesh()->triangles;
+    }
+
+    ITMMesh::Triangle* triangleArray = cpu_triangles->GetData(MEMORYDEVICE_CPU);
+
+    complete_point_cloud.width =
+        main_engine_->GetMesh()->noTotalTriangles * 3; // Point cloud has 3 points per triangle.
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: Is this true?
Are there no duplicate points?

```
*-----*
| \ / |
*--*--*
```

3 triangles, 5 points.

File: InfiniTAM/Engine/RosEngine.cpp [hunk 12]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+      complete_point_cloud.points[i + 1].x = triangleArray[i].p1.x;
+      complete_point_cloud.points[i + 1].y = triangleArray[i].p1.y;
+      complete_point_cloud.points[i + 1].z = triangleArray[i].p1.z;
+
+      complete_point_cloud.points[i + 2].x = triangleArray[i].p2.x;
+      complete_point_cloud.points[i + 2].y = triangleArray[i].p2.y;
+      complete_point_cloud.points[i + 2].z = triangleArray[i].p2.z;
+
+    }
+
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp:215]: make sure you delete duplicate points.

File: InfiniTAM/Engine/RosEngine.cpp [hunk 13]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+    if (shoulDelete) {
+      delete cpu_triangles;
+    }
+
+    pcl::toROSMsg(complete_point_cloud, complete_point_cloud2);
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: move this 
```C++
sensor_msgs::PointCloud2 complete_point_cloud2;
```
here, as you are not using it before. (and rename the variable ;))

File: InfiniTAM/Engine/RosEngine.cpp [hunk 14]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+    complete_point_cloud2.header.frame_id = camera_frame_id_;
+    return complete_point_cloud2;
+
+  } else {
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: you can get rid of the else case if you hard check this in the beginning.

File: InfiniTAM/Engine/RosEngine.cpp [hunk 15]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
 Vector2i RosEngine::getDepthImageSize(void) { return image_size_depth_; }
 Vector2i RosEngine::getRGBImageSize(void) { return image_size_rgb_; }
 
+sensor_msgs::PointCloud2 RosEngine::conversionToPCL() {
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: Can you call this method `extractMeshToPcl(pcl::PointCloud<pcl::PointXYZ>::Ptr out_cloud)` and then assign this to a sensor_msgs::PointCloud2 message outside, in the service callback.

File: InfiniTAM/Engine/RosEngine.cpp [hunk 16]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+    complete_point_cloud.width =
+        main_engine_->GetMesh()->noTotalTriangles * 3; // Point cloud has 3 points per triangle.
+    complete_point_cloud.height =
+        1;  // TODO(gocarlos) how to convert to structured point cloud?
+    complete_point_cloud.is_dense = false;
+    complete_point_cloud.points.resize(complete_point_cloud.width *
+                                       complete_point_cloud.height);
+
+    ROS_INFO_STREAM("this_mesh->noTotalTriangles: "
+                    << main_engine_->GetMesh()->noTotalTriangles);
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp]: make a proper sentence.

File: InfiniTAM/Engine/RosEngine.cpp [hunk 17]
```diff
@@ -156,6 +165,83 @@ bool RosEngine::hasMoreImages(void) {
...
+
+  } else {
+    ROS_ERROR("main_engine_ is NULL");
+  }
+  return complete_point_cloud2;
+}
+
+bool RosEngine::PublishMap(std_srvs::Empty::Request& request,
+                           std_srvs::Empty::Response& response) {
+
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.cpp:234]: do the conversion to `sensor_msgs::PointCloud2` here.

File: InfiniTAM/Engine/RosEngine.h [hunk 18]
```diff
@@ -2,10 +2,11 @@
 
 #pragma once
 
+#include <glog/logging.h>
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.h]: follow the include ordering google coding style

File: InfiniTAM/Engine/RosEngine.h [hunk 19]
```diff
@@ -50,8 +66,16 @@ class RosEngine : public ImageSourceEngine {
...
+  sensor_msgs::PointCloud2 conversionToPCL(void);
+
+  // ROS Service Callback method which published the mesh as PointCloud
+  bool PublishMap(std_srvs::Empty::Request& request,
+                  std_srvs::Empty::Response& response);
 };
-}
-}
+}  // namespace Engine
+}  // namespace InfiniTAM
```
Comment (ffurrer) [inline InfiniTAM/Engine/RosEngine.h:82]: +1

File: InfiniTAM/ITMLib/CMakeLists.txt [hunk 20]
```diff
@@ -12,6 +12,7 @@ SET(ITMLIB_ENGINE_SOURCES
 Engine/ITMColorTracker.cpp
 Engine/ITMDenseMapper.cpp
 Engine/ITMDepthTracker.cpp
+Engine/ITMExternalTracker.cpp #
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/CMakeLists.txt]: remove `#` here and below, unless they still have some meaning. (If they are not needed, remove the line and the corresponding file).

File: InfiniTAM/ITMLib/Engine/DeviceSpecific/CPU/ITMExternalTracker_CPU.cpp [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+ITMExternalTracker_CPU::ITMExternalTracker_CPU(
+    Vector2i imgSize, TrackerIterationType* trackingRegime,
+    int noHierarchyLevels, const ITMLowLevelEngine* lowLevelEngine)
+    : ITMExternalTracker(imgSize, trackingRegime, noHierarchyLevels,
+                         lowLevelEngine, MEMORYDEVICE_CPU) {}
+
+ITMExternalTracker_CPU::~ITMExternalTracker_CPU(void) {}
+
+void ITMExternalTracker_CPU::F_oneLevel(float* f, ITMPose* pose) {
+//  int noTotalPoints = trackingState->pointCloud->noTotalPoints;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/DeviceSpecific/CPU/ITMExternalTracker_CPU.cpp]: what about these commented lines?
Either remove them, implement them, or remove the file.

File: InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+//    int noTotalPoints, Matrix4f M, Vector4f projParams, Vector2i imgSize);
+//
+//__global__ void externalTrackerOneLevel_g_rt_device(
+//    float* g_out, float* h_out, Vector4f* locations, Vector4f* colours,
+//    Vector4s* gx, Vector4s* gy, Vector4u* rgb, int noTotalPoints, Matrix4f M,
+//    Vector4f projParams, Vector2i imgSize);
+//__global__ void externalTrackerOneLevel_g_ro_device(
+//    float* g_out, float* h_out, Vector4f* locations, Vector4f* colours,
+//    Vector4s* gx, Vector4s* gy, Vector4u* rgb, int noTotalPoints, Matrix4f M,
+//    Vector4f projParams, Vector2i imgSize);
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu]: same, if these are not needed remove them... (you can also say that you are not supporting GPUs), or add some
```C++
LOG(FATAL) << "Not implemented.";
```
to dummy functions.

File: InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu [hunk 23]
```diff
@@ -0,0 +1,187 @@
...
+//  ITMSafeCall(cudaMalloc((void**)&f_device,
+//                         sizeof(Vector2f) * (imgSize.x * imgSize.y / 128)));
+//  ITMSafeCall(cudaMalloc((void**)&g_device, sizeof(float) * dim_g *
+//                                                (imgSize.x * imgSize.y / 128)));
+//  ITMSafeCall(cudaMalloc((void**)&h_device, sizeof(float) * dim_h *
+//                                                (imgSize.x * imgSize.y / 128)));
+//
+//  f_host = new Vector2f[imgSize.x * imgSize.y / 128];
+//  g_host = new float[dim_g * imgSize.x * imgSize.y / 128];
+//  h_host = new float[dim_h * imgSize.x * imgSize.y / 128];
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu]: same

File: InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu [hunk 24]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+ITMExternalTracker_CUDA::~ITMExternalTracker_CUDA(void) {
+//  ITMSafeCall(cudaFree(f_device));
+//  ITMSafeCall(cudaFree(g_device));
+//  ITMSafeCall(cudaFree(h_device));
+//
+//  delete[] f_host;
+//  delete[] g_host;
+//  delete[] h_host;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu]: same

File: InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+void ITMExternalTracker_CUDA::F_oneLevel(float* f, ITMPose* pose) {
+//  int noTotalPoints = trackingState->pointCloud->noTotalPoints;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu]: same

File: InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+//  } else {
+//    scaleForOcclusions = (float)noTotalPoints / countedPoints_valid;
+//  }
+//
+//  f[0] = final_f * scaleForOcclusions;
+}
+
+void ITMExternalTracker_CUDA::G_oneLevel(float* gradient, float* hessian,
+                                         ITMPose* pose) const {
+//  int noTotalPoints = trackingState->pointCloud->noTotalPoints;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/DeviceSpecific/CUDA/ITMExternalTracker_CUDA.cu]: same

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.cpp [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+ITMExternalTracker::ITMExternalTracker(Vector2i imgSize,
+                                       TrackerIterationType* trackingRegime,
+                                       int noHierarchyLevels,
+                                       const ITMLowLevelEngine* lowLevelEngine,
+                                       MemoryDeviceType memoryType) {
+  viewHierarchy = new ITMImageHierarchy<ITMViewHierarchyLevel>(
+      imgSize, trackingRegime, noHierarchyLevels, memoryType);
+
+  this->lowLevelEngine = lowLevelEngine;
+  ROS_INFO("ITMExternalTracker");
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.cpp]: don't use ROS here, if you really need output here, then use, e.g.:
```C++
LOG(INFO) << "Using an ITMExternalTracker.";
```

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.cpp [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+ITMExternalTracker::~ITMExternalTracker(void) {
+//  delete viewHierarchy;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.cpp]: rm?

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.cpp [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+
+
+//  ITMPose currentPara(
+//      view->calib->trafo_rgb_to_depth.calib_inv
+//          * trackingState->pose_d->GetM());
+
+  // these following will coerce the result back into the chosen
+  // parameterization for rotations
+//  trackingState->pose_d->SetM(
+//      view->calib->trafo_rgb_to_depth.calib * currentPara.GetM());
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.cpp]: rm

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.h [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+  const ITMLowLevelEngine* lowLevelEngine;
+
+  void PrepareForEvaluation(const ITMView* view);
+
+ protected:
+  TrackerIterationType iterationType;
+
+  // write into the pose
+  //  trackingState->pose_d;
+  // TODO(gocarlos) change this here, provide a first guess from the ROS TF
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.h]: ?

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.h [hunk 31]
```diff
@@ -0,0 +1,104 @@
+// Copyright 2014-2015 Isis Innovation Limited and the authors of InfiniTAM
+
+#pragma once
+
+#include <ros/ros.h>
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.h]: try not to use ros here.

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.h [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+#include <tf/transform_listener.h>
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.h]: the tf_listener should be in the rosengine.

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.h [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+  int levelId;
+
+  int countedPoints_valid;
+
+ public:
+  class EvaluationPoint {
+   public:
+    float f(void) { return cacheF; }
+    const float* nabla_f(void) {
+      if (cacheNabla == NULL) computeGradients(false);
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.h]: nullptr

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.h [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+   the external tracking system and get the current view frame from it.
+*/
+class ITMExternalTracker : public ITMTracker {
+ private:
+  const ITMLowLevelEngine* lowLevelEngine;
+
+  void PrepareForEvaluation(const ITMView* view);
+
+ protected:
+  TrackerIterationType iterationType;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.h]: use our or their variable naming conventions but don't mix.

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.h [hunk 35]
```diff
@@ -0,0 +1,104 @@
...
+    float cacheF;
+    float* cacheNabla;
+    float* cacheHessian;
+  };
+
+  EvaluationPoint* evaluateAt(ITMPose* para) const {
+    return new EvaluationPoint(para, this);
+  }
+
+
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.h]: rm vspaces

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.h [hunk 36]
```diff
@@ -0,0 +1,104 @@
...
+    ITMPose* mPara;
+    const ITMExternalTracker* mParent;
+
+    float cacheF;
+    float* cacheNabla;
+    float* cacheHessian;
+  };
+
+  EvaluationPoint* evaluateAt(ITMPose* para) const {
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.h]: do you need this?

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.h [hunk 37]
```diff
@@ -0,0 +1,104 @@
...
+    return new EvaluationPoint(para, this);
+  }
+
+
+
+  int numParameters(void) const {
+    return (iterationType == TRACKER_ITERATION_ROTATION) ? 3 : 6;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMExternalTracker.h]: define `constexpr kSomeGoodConstantVariableName = ...`.

File: InfiniTAM/ITMLib/Engine/ITMMainEngine.h [hunk 38]
```diff
@@ -84,6 +84,9 @@ namespace ITMLib
 				InfiniTAM_IMAGE_UNKNOWN
 			};
 
+      /// Gives access to the meshing engine, is needed to get a full mesh.
+      ITMMeshingEngine<ITMVoxel, ITMVoxelIndex> * GetMeshingEngine(void) { return meshingEngine; }
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMMainEngine.h:87]: keep indentation style.

If you want to fix it for the whole file, do it in a separate PR.

File: InfiniTAM/ITMLib/Objects/ITMMesh.h [hunk 39]
```diff
@@ -44,6 +44,7 @@ namespace ITMLib
 
 				Triangle *triangleArray = cpu_triangles->GetData(MEMORYDEVICE_CPU);
 
+				// TODO(gocarlos): build pcl out oftriangleArray those points
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Objects/ITMMesh.h]: ?

File: InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp [hunk 40]
```diff
@@ -7,9 +7,9 @@
 using namespace ITMLib::Objects;
 
 ITMLibSettings::ITMLibSettings(void)
-    : sceneParams(0.02f, 100, 0.005f, 0.2f, 3.0f, false) {
+    : sceneParams(0.02f, 100, 0.008f, 0.2f, 3.0f, false) {
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp]: don't change these default values, you can do change them locally.

File: InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp [hunk 41]
```diff
@@ -7,9 +7,9 @@
...
   /// depth threashold for the ICP tracker
-  depthTrackerICPThreshold = 0.1f * 0.1f;
+  depthTrackerICPThreshold = 0.01f;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp]: same, was probably indicating that this is some squared icp distance threshold.

File: InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp [hunk 42]
```diff
@@ -27,7 +27,9 @@ ITMLibSettings::ITMLibSettings(void)
 #endif
 #endif
 
-  // deviceType = DEVICE_CPU;
+  // currently using only CPU for External tracker
+  // CPU is very slow.
+//  deviceType = DEVICE_CPU;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp]: undo

File: InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp [hunk 43]
```diff
@@ -39,12 +41,12 @@ ITMLibSettings::ITMLibSettings(void)
   /// enable or disable bilateral depth filtering;
   useBilateralFilter = false;
 
-  // trackerType = TRACKER_COLOR;
+//   trackerType = TRACKER_COLOR;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp]: undo

File: InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp [hunk 44]
```diff
@@ -39,12 +41,12 @@ ITMLibSettings::ITMLibSettings(void)
...
   trackerType = TRACKER_ICP;
   // trackerType = TRACKER_REN;
   // trackerType = TRACKER_IMU;
   // trackerType = TRACKER_WICP;
-  // trackerType = TRACKER_EXTERNAL;
+//    trackerType = TRACKER_EXTERNAL;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp]: undo

File: InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp [hunk 45]
```diff
@@ -81,6 +83,10 @@ ITMLibSettings::ITMLibSettings(void)
     printf(
         "Error: Color tracker requires a voxel type with color information!\n");
   }
+  if (trackerType == TRACKER_EXTERNAL) {
+    // TODO(gocarlos) Do some stuff here.
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp]: so do you need to do stuff there?

File: InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp [hunk 46]
```diff
@@ -81,6 +83,10 @@ ITMLibSettings::ITMLibSettings(void)
...
+    std::cout << "TRACKER_EXTERNAL" << std::endl;
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Utils/ITMLibSettings.cpp]: use `LOG(INFO) << ...`

File: InfiniTAM/Engine/ImageSourceEngine.h [hunk 47]
```diff
@@ -9,14 +9,17 @@ namespace Engine {
...
   virtual ~ImageSourceEngine() {}
 
   virtual bool hasMoreImages(void) = 0;
   virtual void getImages(ITMUChar4Image* rgb, ITMShortImage* rawDepth) = 0;
   virtual Vector2i getDepthImageSize(void) = 0;
   virtual Vector2i getRGBImageSize(void) = 0;
+
+  // Little hack in order to access
+  // the main_engine in the RosEngine (RosEngine inherits from ImageSourceEngine)
+  ITMMainEngine* main_engine_ = NULL;
```
Comment (ffurrer) [inline InfiniTAM/Engine/ImageSourceEngine.h:22]: Is there no nicer way to do this?

**Review by ffurrer — CHANGES_REQUESTED** [review]

File: InfiniTAM/Engine/RosEngine.cpp (hunk 11 above)
Comment (gocarlos) [reply InfiniTAM/Engine/RosEngine.cpp]: Thanks, I meant the triangle has at least 3 points, not the point cloud. Now what do you advise, check if there are duplicates in the point cloud (cpu expensive?) or let the point cloud through a voxel grid filter at the end of the method? 

I tried to figure out how STL files are build and if PCL has a method to transform from mesh to pcl. They have but as far as I can see they also do not count for the fact that there could be duplicates in the point cloud (https://github.com/PointCloudLibrary/pcl/blob/46cb8fe5589e88e36d79f9b8b8e5f4ff4fceb5de/io/src/vtk_lib_io.cpp )

File: InfiniTAM/ITMLib/Engine/ITMExternalTracker.h (hunk 34 above)
Comment (gocarlos) [reply InfiniTAM/ITMLib/Engine/ITMExternalTracker.h]: Those variable names were already here, it appears that the authors did not use the same rules all the time. Sometimes they use camelCase, sometimes CamelCase and sometimes camel_case ;)

File: InfiniTAM/InfiniTAM_node.cpp [hunk 48]
```diff
@@ -124,6 +132,13 @@ int main(int argc, char** argv) try {
   ros::init(argc, argv, "infinitam_node");
   ros::NodeHandle node_handle("~");
 
+  ITMMainEngine* main_engine = NULL;
```
Comment (ffurrer) [inline InfiniTAM/InfiniTAM_node.cpp:134]: use `nullptr` instead of `NULL` (see e.g. http://stackoverflow.com/questions/13816385/what-are-the-advantages-of-using-nullptr)

File: InfiniTAM/InfiniTAM_node.cpp [hunk 49]
```diff
@@ -96,7 +100,11 @@ static void CreateDefaultImageSource(const char* arg1, const char* arg2,
   }
   if (image_source == NULL) {
     printf("Checking if there are suitable ROS messages being published.\n");
+
     image_source = new RosEngine(node_handle, calibration_filename);
+
```
Comment (ffurrer) [inline InfiniTAM/InfiniTAM_node.cpp:105]: remove one vspace

File: InfiniTAM/ITMLib/Engine/ITMTrackerFactory.h [hunk 50]
```diff
@@ -55,7 +55,7 @@ namespace ITMLib
       {
         makers.insert(std::make_pair(ITMLibSettings::TRACKER_COLOR, &MakeColourTracker));
         makers.insert(std::make_pair(ITMLibSettings::TRACKER_ICP, &MakeICPTracker));
-		makers.insert(std::make_pair(ITMLibSettings::TRACKER_WICP, &MakeWeightedICPTracker));
+    makers.insert(std::make_pair(ITMLibSettings::TRACKER_WICP, &MakeWeightedICPTracker));
```
Comment (ffurrer) [inline InfiniTAM/ITMLib/Engine/ITMTrackerFactory.h]: alignment is still somewhat wrong ;)

File: InfiniTAM/InfiniTAM_node.cpp [hunk 51]
```diff
@@ -169,24 +184,24 @@ int main(int argc, char** argv) try {
   }
 
   printf("initialising ...\n");
-  ImageSourceEngine* image_source = NULL;
-  IMUSourceEngine* imu_source = NULL;
 
-  CreateDefaultImageSource(arg1, arg2, arg3, arg4, node_handle, image_source,
-                           imu_source);
+  SetUpSources(arg1, arg2, arg3, arg4, node_handle, image_source, imu_source);
```
Comment (ffurrer) [inline InfiniTAM/InfiniTAM_node.cpp:187]: why are you not passing the main_engine? So you could avoid having 
```C++
ITMMainEngine* main_engine_ = NULL;
```
in the `ImageSourceEngine`, I guess.

File: InfiniTAM/InfiniTAM_node.cpp [hunk 52]
```diff
@@ -169,24 +184,24 @@ int main(int argc, char** argv) try {
...
 
-  ITMLibSettings* internal_settings = new ITMLibSettings();
-  ITMMainEngine* main_engine = new ITMMainEngine(
-      internal_settings, &image_source->calib, image_source->getRGBImageSize(),
-      image_source->getDepthImageSize());
+  main_engine = new ITMMainEngine(internal_settings, &image_source->calib,
+                                  image_source->getRGBImageSize(),
+                                  image_source->getDepthImageSize());
+
+  image_source->main_engine_ = main_engine;
```
Comment (ffurrer) [inline InfiniTAM/InfiniTAM_node.cpp:198]: and you wouldn't need to set this.

**Review by ffurrer — COMMENTED** [review]
some more small modifications, other than that lgtm

File: InfiniTAM/InfiniTAM_node.cpp (hunk 51 above)
Comment (gocarlos) [reply InfiniTAM/InfiniTAM_node.cpp:187]: I'm not passing the main_engine because main_engine has to be initialized after the initialization of image_source (main_engine gets an instance of ImageSourceEngine and RosEngine inherits from ImageSourceEngine).
And in RosEngine I need to access the ITMMainEngine, its a chicken or the egg dilemma.

Comment (gocarlos) [reply InfiniTAM/InfiniTAM_node.cpp:187]: Eventually we can make it better in the new ros node in the branch external_tracker
